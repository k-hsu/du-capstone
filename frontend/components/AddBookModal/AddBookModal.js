import React, { useState } from "react";
import { useForm } from "react-hook-form";
import Form from "../Form/Form";
import Input from "../Input/Input";
import Modal from "../Modal/Modal";
import TextArea from "../TextArea/TextArea";

const AddBookModal = ({ onClose, onSubmit }) => {
  const [unloaded, setUnloaded] = useState(true);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({ shouldUnregister: true });

  const onSubmitForm = (results) => {
    onSubmit(results);
  };

  const { ref, ...titleRegister } = register("title", {
    required: "title is required",
  });

  return (
    <Modal
      title="Add New Book"
      submitText="Add Book"
      onClose={(e) => {
        if (!e.key || e.key === "Enter") {
          onClose();
        }
      }}
      onSubmit={(e) => {
        if (!e.key || e.key === "Enter") {
          handleSubmit(onSubmitForm)(e);
        }
      }}
    >
      <Form>
        <Input
          labelText="Title"
          errorMessage={errors.title?.message}
          ref={(innerRef) => {
            if (unloaded && innerRef) {
              innerRef.focus();
              setUnloaded(false);
            }
            return ref(innerRef);
          }}
          {...titleRegister}
        />
        <Input
          labelText="Author"
          errorMessage={errors.author?.message}
          {...register("author", {
            required: "author is required",
            validate: (value) =>
              value.trim().split(" ").length === 2 ||
              "author must have a first name and last name",
          })}
        />
        <TextArea
          rows="6"
          labelText="Description"
          errorMessage={errors.description?.message}
          {...register("description")}
        />
      </Form>
    </Modal>
  );
};

export default AddBookModal;
