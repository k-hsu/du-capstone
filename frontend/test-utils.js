import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderHook } from "@testing-library/react-hooks";

function setup(jsx) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

export * from "@testing-library/react";
export { renderHook, setup };
