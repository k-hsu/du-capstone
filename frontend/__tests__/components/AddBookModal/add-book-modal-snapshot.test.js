import React from "react";
import { renderer } from "../../../test-utils";
import AddBookModal from "../../../components/AddBookModal/AddBookModal";

jest.mock("react-dom", () => {
  const originalModule = jest.requireActual("react-dom");

  return {
    ...originalModule,
    createPortal: (element) => element,
  };
});

describe("AddBookModal Snapshot", () => {
  const spy = jest.spyOn(document, "getElementById");
  spy.mockReturnValue(<div id="app-root" />);
  it("should render modal component styles", () => {
    const component = renderer.create(<AddBookModal />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
