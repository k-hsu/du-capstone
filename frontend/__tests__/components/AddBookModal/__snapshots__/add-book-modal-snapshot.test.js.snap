// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBookModal Snapshot should render modal component styles 1`] = `
<div
  className="css-2wtgp9"
  onClick={[Function]}
>
  <div
    className="css-15e9n3k"
    onClick={[Function]}
  >
    <div
      className="css-1pgkpxz"
      direction="row"
    >
      <p
        className="css-dtk8j7"
        fontFamily="IBM Plex Sans"
        fontWeight="bold"
      >
        Add New Book
      </p>
      <p
        className="css-1lkb4ir"
        cursor="pointer"
        fontFamily="IBM Plex Sans"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        X
      </p>
    </div>
    <form
      className="css-1v885rd"
    >
      <label
        className="css-1iyoj2o"
      >
        <p
          className="css-16kvdet"
          fontFamily="IBM Plex Sans"
          fontWeight="700"
        >
          Title
        </p>
        <input
          className="css-mn8zue"
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </label>
      <label
        className="css-1iyoj2o"
      >
        <p
          className="css-16kvdet"
          fontFamily="IBM Plex Sans"
          fontWeight="700"
        >
          Author
        </p>
        <input
          className="css-mn8zue"
          name="author"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </label>
      <label
        className="css-1iyoj2o"
      >
        <p
          className="css-16kvdet"
          fontFamily="IBM Plex Sans"
          fontWeight="700"
        >
          Description
        </p>
        <textarea
          className="css-mn8zue"
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          rows="6"
        />
      </label>
    </form>
    <div
      className="css-1a6mal2"
      direction="row"
    >
      <p
        className="css-1fd35ut"
        color="#F00"
        cursor="pointer"
        fontFamily="IBM Plex Sans"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        Cancel
      </p>
      <button
        className="css-wkrzoz"
        onClick={[Function]}
      >
        <p
          className="css-1sofp4k"
          fontFamily="IBM Plex Sans"
        >
          Add Book
        </p>
      </button>
    </div>
  </div>
</div>
`;
